src/Sylius/Bundle/ApiBundle/DataProvider/ShippingMethodsSubresourceDataProvider.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\DataProvider;
+
+use ApiPlatform\Core\DataProvider\RestrictedDataProviderInterface;
+use ApiPlatform\Core\DataProvider\SubresourceDataProviderInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\ShipmentInterface;
+use Sylius\Component\Core\Model\ShippingMethodInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Sylius\Component\Core\Repository\ShipmentRepositoryInterface;
+use Sylius\Component\Registry\ServiceRegistryInterface;
+use Sylius\Component\Shipping\Resolver\CompositeMethodsResolver;
+use Webmozart\Assert\Assert;
+
+final class ShippingMethodsSubresourceDataProvider implements RestrictedDataProviderInterface, SubresourceDataProviderInterface
+{
+    /** @var OrderRepositoryInterface */
+    private $orderRepository;
+
+    /** @var ShipmentRepositoryInterface */
+    private $shipmentRepository;
+
+    /** @var CompositeMethodsResolver */
+    private $compositeMethodsResolver;
+
+    /** @var ServiceRegistryInterface */
+    private $calculators;
+
+    public function __construct(OrderRepositoryInterface $orderRepository, ShipmentRepositoryInterface $shipmentRepository, CompositeMethodsResolver $compositeMethodsResolver, ServiceRegistryInterface $calculators)
+    {
+        $this->orderRepository = $orderRepository;
+        $this->shipmentRepository = $shipmentRepository;
+        $this->compositeMethodsResolver = $compositeMethodsResolver;
+        $this->calculators = $calculators;
+    }
+
+    public function getSubresource(string $resourceClass, array $identifiers, array $context, string $operationName = null)
+    {
+        $subresourceIdentifiers = $context['subresource_identifiers'];
+
+        /** @var OrderInterface|null $cart */
+        $cart = $this->orderRepository->findCartByTokenValue($subresourceIdentifiers['id']);
+        Assert::notNull($cart);
+
+        /** @var ShipmentInterface $shipment */
+        $shipment = $this->shipmentRepository->findOneBy(['id' => $subresourceIdentifiers['shipments']]);
+        Assert::notNull($shipment);
+
+        Assert::true($cart->hasShipment($shipment), 'Shipment doesn\'t match for order');
+
+        return $this->getCheckoutShippingMethods($cart, $shipment);
+    }
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
+    {
+        return
+            is_a($resourceClass, ShippingMethodInterface::class, true) &&
+            !isset($subresourceIdentifiers['id']) &&
+            !isset($subresourceIdentifiers['shipments'])
+        ;
```

Comment:
This condition allows using this SubresourceDataProvider only for `/new-api/orders/{id}/shipments/{shipments}/methods` endpoint